### PR TITLE
added tcp keep alive

### PR DESCRIPTION
**Pull request by syntacs**
always on, hard coded times

shall I create additional parameters for client/server in order to make it configurable ?

If yes, what shall be configurable ?

### CODE REVIEW

**Comment by jchristn** [comment]
Hi @syntacs thanks for submitting this.  Help me understand the use case here - the connections stay alive already, so I'm not sure I understand the context of the PR.

**Comment by DuAell** [comment]
If a client disconnects, is the server aware of it ?
If the server disconnects, are the clients aware of it ?
What if the client/server is stuck for some reason and stay connected but cannot send/receive messages ?

**Comment by jchristn** [comment]
Hi @DuAell 

1) If a client disconnects, is the server aware of it ?
2) If the server disconnects, are the clients aware of it ?

Yes.  Please refer to the bottom of the README for the cases for which I've optimized.  I don't have an objection to adding TCP keepalives, but they don't add value in terms of disconnect detection beyond what's already implemented.

## Disconnection Handling

The project TcpTest (https://github.com/jchristn/TcpTest) was built specifically to provide a reference for WatsonTcp to handle a variety of disconnection scenarios.  These include:

| Test case | Description | Pass/Fail |
|---|---|---|
| Server-side dispose | Graceful termination of all client connections | PASS |
| Server-side client removal | Graceful termination of a single client | PASS |
| Server-side termination | Abrupt termination due to process abort or CTRL-C | PASS |
| Client-side dispose | Graceful termination of a client connection | PASS |
| Client-side termination | Abrupt termination due to a process abort or CTRL-C | PASS |

3) What if the client/server is stuck for some reason and stay connected but cannot send/receive messages ?

In that case a keepalive would simply keep the connection open, which is the opposite of what you would want to have happen.

My recommendation would be to periodically send *application-layer heartbeats*.  TCP keepalives will keep the socket alive *underneath the application* but for this exact use case (client/server stuck) TCP could remain alive but one endpoint might not be communicative.  Something built into the application to periodically validate the connection would be of value here.

**Comment by syntacs** [comment]
tcp keep alive detects tcp communication partners who vanish all of a sudden, e.g. because a network cable was pulled. It might also happen that, after a long period of inactivity, a switch 'forgets' a connected tcp endpoint.

If e.g. a client vanishes all of a sudden, the server will not detect this as long as the server does not try to send to the vanished client (which in many case does not happen)

All resources will stay allocated inside the server, eventually gathering up until the connection limit is reached.

Sample scenario:

- establish a client - server connection

- pull network cable

- terminate client

- re insert network cable

- restart client -> server will have a second connection, the first one still 'open'. Without keep alive, the dead first connection will stay forever.


https://en.wikipedia.org/wiki/Keepalive


btw.

many keep alive explanations consider client termination through task manager a 'sudden vanish'. This is not correct (might have been different in the past). Upon forceful process termination, windows sends a RST to all active tcp connections -> the server would correctly detect this case.

compared to application level self implemented alive strategies, keep alives on tcp level has the nice feature to be active also while a process is interrupted inside a debugger. Connections will not break because a process is being debugged.

File: WatsonTcp/WatsonTcpClient.cs [hunk 1]
```diff
@@ -375,6 +375,8 @@ public void Start()
                 Logger?.Invoke("[WatsonTcpClient] Connecting to " + _ServerIp + ":" + _ServerPort);
 
                 _Client.LingerState = new LingerOption(true, 0);
+                byte[] outValue = new byte[10];
+                _Client.Client.IOControl(IOControlCode.KeepAliveValues, new KeepAliveValues().Values, outValue);
```
Comment (Job79) [inline WatsonTcp/WatsonTcpClient.cs:379]: This is windows specific and throws an exception on all other systems.

File: WatsonTcp/KeepAliveValues.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            onoff = 1;
+            keepalivetime = 10000;
+            keepaliveinterval = 1000;
+        }
+        public byte[] Values
+        {
+            get
+            {
+                MemoryStream bytes = new MemoryStream();
```
Comment (Job79) [inline WatsonTcp/KeepAliveValues.cs:24]: Probably a memory leak

File: WatsonTcp/WatsonTcpClient.cs (hunk 1 above)
Comment (Job79) [reply WatsonTcp/WatsonTcpClient.cs:379]:
```cs
socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 5); // 5 seconds
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 5); // 5 seconds
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount, 1); 
```
This is supposed to do the same and works for me (linux 5.8, dotnet 3.1)
Does this also work on windows?

**Comment by jchristn** [comment]
Hi @syntacs - thanks; the network cable pull use case is one that should be addressed.  I'm familiar with TCP keepalives but hadn't considered the impact of a switch freezing or a cable being pulled (and packets dropping on the floor).

I love your idea of a separate keepalive value class and will test using values stored there with the sample code from @Job79 on .NET Framework/.NET Core on both Windows and Ubuntu.  I may also scope (and include) an application-layer mechanism. 
 Hopefully will have a commit this weekend.

Cheers, Joel

**Comment by jchristn** [comment]
Hi @Job79 @syntacs @DuAell bit of a snag here, the ```SocketOptionName``` values for TCP keepalives are *NOT* present in ```netstandard2.0``` or ```netstandard2.1```.  See: https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketoptionname?view=netstandard-2.1

There is an open issue on this: https://github.com/dotnet/standard/issues/1474

Nor are they present in *ANY* version of .NET Framework that I could find: https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketoptionname?view=netframework-4.8

Is someone aware of an approach that would support both .NET Standard and .NET Framework?

**Comment by jchristn** [comment]
It looks like I can use this, as it works for .NET Standard and .NET Framework both: https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.settcpkeepalive

This does not however allow you to specify the retry count, which other implementations provide.

Thoughts?

**Comment by DuAell** [comment]
As stated in the documentation :

> The number of keep-alive probes cannot be changed and is set to 10.

You can also change the interval between each retry.
I think this is sufficient for what we want to achieve.

**Comment by jchristn** [comment]
Thanks @DuAell - I'll finish integration, do some testing, and release.  There will be breaking changes as I've re-organized the settings, events, and callbacks to make the code more tidy.

**Comment by jchristn** [comment]
I have this committed to branch ```v4.3.0-beta``` (see https://github.com/jchristn/WatsonTcp/tree/v4.3.0-beta)

The server seems to detect (after keepalives fail) when the network adapter is disabled (cable unplugged) but no such luck with the client.  It continues to think that the connection is alive.

I've tested both .NET Framework and .NET Core (though I've only tested on Windows).

Would you have a moment to let me know if you have the same results?

**Comment by Job79** [comment]
> I have this committed to branch `v4.3.0-beta` (see https://github.com/jchristn/WatsonTcp/tree/v4.3.0-beta)
> 
> The server seems to detect (after keepalives fail) when the network adapter is disabled (cable unplugged) but no such luck with the client. It continues to think that the connection is alive.
> 
> I've tested both .NET Framework and .NET Core (though I've only tested on Windows).
> 
> Edit:
> 
> I've just tested on Ubuntu 16.04 with .NET Core 3.1 and have the same results. The client never picks up the disconnection, but the server does.
> 
> In my testing it seems that the `ServicePointManager.SetTcpKeepAlive` method doesn't help the client at all, but does work for the server.
> 
> If you are aware of something that will work for both client and server in a multiplatform, multiframework way, please let me know.

I Replaced the socket options inside my test project(not watsontcp) with ServicePointManager.SetTcpKeepAlive, disconnects are no longer detected correctly on both sides. (disconnecting internet cable, LAN)

Isn't the ServicePointManager for http connections only? The documentation of the [ServicePoint](https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepoint?view=netcore-3.1) class: "Provides connection management for HTTP connections." Isn't this also true for the ServicePointManager?

Another solution would be using the socket options for dotnet core, and using IOControl for .net framework.
This can be achieved with preprocessor like this:
```cs
#if NETCOREAPP 
            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 5); 
            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 5);
            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount, 1);
#elif NETFRAMEWORK 
            // Create array with settings
            byte[] keepAlive = new byte[4 * 3];
            // Turn keepalive on
            Buffer.BlockCopy(BitConverter.GetBytes((uint)1), 0, keepAlive, 0, 4);
            // Set amount of time without activity before sending a keepalive to 5 seconds
            Buffer.BlockCopy(BitConverter.GetBytes((uint)5000), 0, keepAlive, 4, 4); 
            // Set keepalive interval to 5 seconds
            Buffer.BlockCopy(BitConverter.GetBytes((uint)5000), 0, keepAlive, 8, 4); 

            // Set the keep-alive settings on the underlying Socket
            socket.IOControl(IOControlCode.KeepAliveValues, keepAlive, null);
#endif
```

Best would be to ship it as an optional function so that watsonTcp automatically behaves the same on all versions. The people using .net framework or .net core would then be able to use this feature. (after adding .net framework and dotnet core to the compile options in the .csproj)

Example syntax:
```cs
public WatsonTcpClient EnableKeepAlive(int aliveTime = 5, int aliveInterval = 5, int retryCount = 1); // <-- function returns WatsonTcpClient

using var client = new WatsonTcpClient().UseKeepAlive();
```

**Comment by jchristn** [comment]
Hi @Job79 I think you're right, it may not be used by general TCP connections.  I was hoping to avoid framework-specific code and preprocessor directives but you're probably right, it's the way I would need to go forward.  Thanks for the suggestion and sample code.

**Comment by jchristn** [comment]
Hi all, quick update.  I moved to framework-specific code (thank you @Job79 for the example) and tested again.  The server detects disconnection as expected, but the client does not.  The latest is in the ```v4.3.0-beta``` branch.  I plan to release it tomorrow unless anyone has any concerns.  Cheers, Joel

**Comment by jchristn** [comment]
Hi all, I pushed v4.3.0 to NuGet this afternoon (https://www.nuget.org/packages/WatsonTcp/4.3.0) and it should be available in the next few minutes.  Please note there are breaking changes.  Please re-open if you find issues or enhancements.  Thanks for everyone's help on this one; I'll make sure it propagates to my other TCP projects.

**Pull request closed without merging**